crates/project_panel/src/project_panel.rs

```diff
@@ -1833,6 +1932,50 @@ impl ProjectPanel {
         None
     }
 
+    fn sanitized_marked_entries(&self, cx: &AppContext) -> BTreeSet<SelectedEntry> {
+        let marked = self.marked_entries();
+        if marked.is_empty() {
+            return BTreeSet::default();
+        }
+
+        let project = self.project.read(cx);
+        let entries_by_worktree: HashMap<WorktreeId, Vec<SelectedEntry>> =
+            marked
+                .into_iter()
+                .fold(HashMap::default(), |mut map, entry| {
+                    map.entry(entry.worktree_id).or_default().push(entry);
+                    map
+                });
+
+        let mut result = BTreeSet::new();
+        for (worktree_id, entries) in entries_by_worktree {
+            if let Some(worktree) = project.worktree_for_id(worktree_id, cx) {
+                let worktree = worktree.read(cx);
+
+                let marked_paths: HashSet<Arc<Path>> = entries
+                    .iter()
+                    .filter_map(|entry| worktree.entry_for_id(entry.entry_id))
```

Comment:
My spider-sense hint that we can do better and this is dangerously close to something quadratic.
I do not have good ideas on how to fix this, but a few obvious things:
* we do not need add any files here, only dirs can exclude something
* then we can ask during filtering, if filtered entry's path `starts_with && not equal` an entry from `marked_paths`
* I wonder if ordering these `marked_paths` makes any sense, as this way we'll get the "starts_with" triggering faster (if ordered right)
* NIT, but naming could be improved: `marked_paths` (those are something else), `result` (more like `sanitized_paths`) and even `sanitized_marked_entries` (more like `disjoint_paths_for_removal`) could be named better
* I feel there's a better way to check such things in general, but not insisting